packages/ai/core/tool/mcp/client.ts

```diff
@@ -0,0 +1,302 @@
+ *
+ * Not supported:
+ * - Client options (e.g. sampling, roots) as they are not needed for tool conversion
+ * - Accepting notifications
+ */
+export class MCPClient {
+  private transport: Transport;
+  private clientInfo: Implementation;
+  private requestMessageId = 0;
+  private responseHandlers: Map<
+    number,
+    (response: JSONRPCResponse | Error) => void
+  > = new Map();
+  private connectionTimeoutMs: number;
+  private requestTimeoutMs: number;
+
+  constructor({
+    transport: transportConfig,
+    name = 'ai-sdk-mcp-client',
+    version = '1.0.0',
+    connectionTimeoutMs = CONNECTION_TIMEOUT_MS,
+    requestTimeoutMs = REQUEST_TIMEOUT_MS,
+  }: MCPClientConfig) {
+    this.transport = createMcpTransport(transportConfig);
+    this._initTransport();
+    this.clientInfo = {
+      name,
+      version,
+    };
+    this.connectionTimeoutMs = connectionTimeoutMs;
+    this.requestTimeoutMs = requestTimeoutMs;
+  }
+
+  async init(): Promise<this> {
+    try {
+      await this.transport.start();
+
+      const result = await this.request({
+        request: {
+          method: 'initialize',
+          params: {
+            protocolVersion: LATEST_PROTOCOL_VERSION,
+            capabilities: {},
+            clientInfo: this.clientInfo,
+          },
+        },
+        resultSchema: InitializeResultSchema,
+        options: { timeout: this.connectionTimeoutMs },
+      });
+
+      if (result === undefined) {
+        throw new AISDKError({
+          name: 'McpClientError',
+          message: 'Server sent invalid initialize result',
+        });
+      }
+
+      if (!SUPPORTED_PROTOCOL_VERSIONS.includes(result.protocolVersion)) {
+        throw new AISDKError({
+          name: 'McpClientError',
+          message: `Server's protocol version is not supported: ${result.protocolVersion}`,
+        });
+      }
+
+      if (!result.capabilities?.tools) {
+        throw new AISDKError({
+          name: 'McpClientError',
+          message: `Server does not support tools`,
+        });
+      }
+
+      // Complete initialization handshake:
+      await this.notification({
+        method: 'notifications/initialized',
+      });
+
+      return this;
+    } catch (error) {
+      void this.close();
+      throw error;
+    }
+  }
+
+  async close(): Promise<void> {
+    await this.transport?.close();
+    this._onclose();
+  }
+
+  async request<T extends ZodType<object>>({
+    request,
+    resultSchema,
+    options,
+  }: {
+    request: Request;
+    resultSchema: T;
+    options?: RequestOptions;
+  }): Promise<z.infer<T>> {
+    return new Promise((resolve, reject) => {
+      options?.signal?.throwIfAborted();
+
+      const timeoutSignal = AbortSignal.timeout(
+        options?.timeout ?? this.requestTimeoutMs,
+      );
+      const signal = options?.signal
+        ? AbortSignal.any([options.signal, timeoutSignal])
+        : timeoutSignal;
```

Comment:
since we pass in an `AbortSignal`, I was thinking we can fully remove the timeout code to keep things simple